crates/log-server/src/rocksdb_logstore/store.rs

```diff
@@ -194,6 +197,145 @@ impl LogStore for RocksDbLogStore {
     ) -> Result<AsyncToken, OperationError> {
         self.writer_handle.enqueue_trim(loglet_id, trim_point).await
     }
+
+    async fn read_records(
+        &mut self,
+        msg: GetRecords,
+        loglet_state: LogletState,
+    ) -> Result<Records, OperationError> {
+        let data_cf = self.data_cf();
+        let loglet_id = msg.loglet_id;
+        // The order of operations is important to remain correct.
+        // The scenario that we want to avoid is the silent data loss. Although the
+        // replicated-loglet reader can ensure contiguous offset range and detect trim-point
+        // detection errors or data loss, we try to structure this to reduce the possibilities for
+        // errors as much as possible.
+        //
+        // If we are reading beyond the tail, the first thing we do is to clip to the
+        // local_tail.
+        let local_tail = loglet_state.local_tail().clone();
+        let trim_point = loglet_state.trim_point();
+
+        let read_from = msg.from_offset.max(trim_point.next());
+        let read_to = msg.to_offset.min(local_tail.offset().prev());
```

Comment:
Shouldn't we have a limit over the size of the queried range ? otherwise a caller can cause a huge spike in memory usage (or crash) if they issued a huge range read.

Reply:
I think this is what is called `size_budget` but no actual limit on the requested size

Reply:
Good call, this can be a nice follow up.